### PR TITLE
[NEXUS-8278] Component/Asset wrappers for OrientVertex

### PR DESCRIPTION
**Pull request by cwilper**
https://issues.sonatype.org/browse/NEXUS-8278

### CODE REVIEW

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import com.tinkerpop.blueprints.impls.orient.OrientVertex;
+import org.joda.time.DateTime;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_ATTRIBUTES;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_LAST_UPDATED;
+
+/**
+ * Wraps an {@code OrientVertex} to provide
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java]: to provide... The suspense is KILLING ME!!! 
:)

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/Asset.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import com.tinkerpop.blueprints.Direction;
+import com.tinkerpop.blueprints.Vertex;
+import com.tinkerpop.blueprints.impls.orient.OrientVertex;
+
+import static org.sonatype.nexus.repository.storage.StorageFacet.E_PART_OF_COMPONENT;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_BLOB_REF;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_CONTENT_TYPE;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_SIZE;
+
+public class Asset
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/Asset.java]: basic javadoc please

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/Component.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import com.tinkerpop.blueprints.impls.orient.OrientVertex;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static org.sonatype.nexus.repository.storage.StorageFacet.E_PART_OF_COMPONENT;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_FORMAT;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_GROUP;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_NAME;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_VERSION;
+
+public class Component
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/Component.java]: basic javadoc please

**Comment by kellyrob99** [comment]
+1

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java (hunk 1 above)
Comment (cwilper) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java]: Heh. I did kind of cut that short eh. Updated.

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/Component.java (hunk 3 above)
Comment (cwilper) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/Component.java]: added basic javadocs in latest (rebased) commit

**Comment by adreghiciu** [comment]
What about reducing visibility of createVertex, deleteVertex, deleteBlob?

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+import static com.google.common.base.Preconditions.checkState;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_ATTRIBUTES;
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_LAST_UPDATED;
+
+/**
+ * Wraps an {@code OrientVertex} to provide a simpler API for working with stored component and asset metadata.
+ */
+public abstract class MetadataNode
+{
+  protected final OrientVertex vertex;
```
Comment (mrprescott) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java]: Now that we're serving up collections of these, I think it's wise to implement equals() and hashCode(), based on referring to the same vertex.

Comment (cwilper) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java]: As these are storage objects, which attributes do you think are important to consider for equality (id only, properties only, or id + properties?)  Note that id is temporary before a record is first committed.

**Comment by cwilper** [comment]
@adreghiciu re: reducing visibility of those methods, I think that's a good goal. currently bucket management works at the OrientVertex level but maybe a simple Bucket wrapper is the way to go. Hmm.

**Comment by cstamas** [comment]
+1

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * Sets a property to a given value, or {@code null} to remove it.
+   */
+  public void set(final String name, final @Nullable Object value) {
+    checkNotNull(name);
+    if (value == null) {
+      vertex.removeProperty(name);
+    }
+    else if (value instanceof DateTime) {
+      vertex.setProperty(name, ((DateTime) value).toDate());
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java]: is there no means to inform orient how to do handle DateTime w/o custom code like this?

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    return null;
+  }
+
+  public String require(final String name) {
+    return require(name, String.class);
+  }
+
+  public <T> T require(final String name, final Class<T> type) {
+    T value = get(name, type);
+    checkState(value != null, "Missing property: " + name);
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java]: use %s form

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/AssetEvent.java [hunk 7]
```diff
@@ -12,17 +12,16 @@
...
  */
 public abstract class AssetEvent
-  extends VertexEvent
 {
-  public AssetEvent(final OrientVertex vertex) {
-    super(vertex);
+  private final Asset asset;
+
+  public AssetEvent(final Asset asset) {
+    this.asset = asset;
```
Comment (jdillon) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/AssetEvent.java]: null check

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java (hunk 5 above)
Comment (cwilper) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/MetadataNode.java]: it does appear there's a way to register a custom serializer/deserializer, but what detail i've found has been incomplete. 

* http://jezzper.com/jezzper/discussions.nsf/0/4B3FB2D9D145E7C9C12579AE0048F722
* https://github.com/dozed/orientdb-test/blob/master/src/main/java/org/noorg/orientdb/test/Database.java#L63

I'll see what I can figure out. Also pinged luca.

**Comment by cwilper** [comment]
http://bamboo.s/browse/NX3-OSSF105-3 :white_check_mark:

File: components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java [hunk 8]
```diff
@@ -200,12 +215,19 @@ private OrientVertex findWithPropertyOwnedBy(String className, String propName,
...
-                                           @Nullable String querySuffix)
+  public Iterable<Asset> findAssets(@Nullable String whereClause,
+                                    @Nullable Map<String, Object> parameters,
+                                    @Nullable Iterable<Repository> repositories,
+                                    @Nullable String querySuffix)
   {
-    return findVertices(V_ASSET, whereClause, parameters, E_OWNS_ASSET, repositories, querySuffix);
+    return Iterables.transform(
+        findVertices(V_ASSET, whereClause, parameters, E_OWNS_ASSET, repositories, querySuffix),
+        new Function<OrientVertex, Asset>() {
```
Comment (kellyrob99) [inline components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/StorageTxImpl.java]: might want to extract the functions used to constants, this one is a repeat but we also shouldn't need to instantiate these more than once

File: plugins/rapture/nexus-coreui-plugin/src/main/java/org/sonatype/nexus/coreui/ComponentComponent.groovy [hunk 9]
```diff
@@ -66,17 +64,17 @@ extends DirectComponentSupport
     }
     StorageTx storageTx = repository.facet(StorageFacet).openTx()
     try {
-      OrientVertex component = storageTx.findComponent(new ORecordId(componentId), storageTx.getBucket())
+      ComponentImpl component = storageTx.findComponent(new ORecordId(componentId), storageTx.getBucket())
```
Comment (kellyrob99) [inline plugins/rapture/nexus-coreui-plugin/src/main/java/org/sonatype/nexus/coreui/ComponentComponent.groovy]: why impl's here instead of interfaces?

**Comment by kellyrob99** [comment]
+1

File: plugins/rapture/nexus-coreui-plugin/src/main/java/org/sonatype/nexus/coreui/ComponentComponent.groovy (hunk 9 above)
Comment (cwilper) [inline plugins/rapture/nexus-coreui-plugin/src/main/java/org/sonatype/nexus/coreui/ComponentComponent.groovy]: D'oh, good catch, will fix. The impls were intentionally package-scoped, but this is groovy code so it escaped my notice. IDEA's "extract interface" is awesome, but does seem to prefer to keep old declarations going at the impl vs the new interface as one might expect.

**Pull request merged**